CoreScriptsRoot/Modules/VR/VirtualKeyboard.lua

```diff
@@ -390,19 +856,74 @@ local function ConstructKeyboardUI(keyboardLayoutDefinition)
 			keyboardOptions.TextBox.Text = newBufferText
 		elseif buffer ~= newBufferText then
 			buffer = newBufferText
-			textEntryField.Text = buffer
+			UpdateTextEntryFieldText(buffer)
+		end
+	end
+
+	local function calculateTextCursorPosition(x, y)
+		x = x - textEntryField.AbsolutePosition.x
+		y = y - textEntryField.AbsolutePosition.y
+
+		for i = 1, #textEntryField.Text do
+			local textSize = TextService:GetTextSize(
+				string.sub(textEntryField.Text, 1, i),
+				ConvertFontSizeEnumToInt(textEntryField.FontSize),
+				textEntryField.Font,
+				textEntryField.AbsoluteSize)
+			if textSize.x > x then
+				return i - 1
+			end
 		end
+
+		return #textEntryField.Text
 	end
 
-	rawset(newKeyboard, "GetKeyByPosition", function(self, x,y)
+	local currentKeyset = nil
+
+	local newCursorPos = nil
+	local myCursorX = nil
+	local myCursorY = nil
+
+	rawset(newKeyboard, "GetCurrentKeyset", function(self)
+		return keysets[currentKeyset]
+	end)
+
+	rawset(newKeyboard, "SetCurrentKeyset", function(self, newKeyset)
+		if newKeyset ~= currentKeyset and keysets[newKeyset] ~= nil then
+			if keysets[currentKeyset] and keysets[currentKeyset].container then
+				keysets[currentKeyset].container.Visible = false
+			end
+
+			currentKeyset = newKeyset
+
+			if keysets[currentKeyset] and keysets[currentKeyset].container then
+				keysets[currentKeyset].container.Visible = true
+			end
+		end
+	end)
+
+	rawset(newKeyboard, "SetVoiceMode", function(self, inVoiceMode)
+		inVoiceMode = inVoiceMode and IsVoiceToTextEnabled()
+
+		local currentKeysetObject = self:GetCurrentKeyset()
+		if currentKeysetObject and currentKeysetObject.container then
+			currentKeysetObject.container.Visible = not inVoiceMode
+		end
+
+		voiceRecognitionContainer.Visible = inVoiceMode
+	end)
+
+	rawset(newKeyboard, "GetKeyByPosition", function(self, x, y)
 		-- There are a lot of keys, we could optimize this by caching some sort of lookup
-		for _, element in pairs(keys) do
-			local minPt = element.AbsolutePosition
-			local maxPt = element.AbsolutePosition + element.AbsoluteSize
-			if minPt.X <= x and maxPt.X >= x and minPt.Y <= y and maxPt.Y >= y then
-				return element
+		local currentKeysetObject = self:GetCurrentKeyset()
+		if currentKeysetObject and currentKeysetObject.keys and currentKeysetObject.container and currentKeysetObject.container.Visible then
+			for _, element in pairs(self:GetCurrentKeyset().keys) do
```

Comment:
minor, but couldn't currentKeysetObject.keys be used for iteration since we got it above?